File: demos/samples/03-agents/python/03_maf_concurrent.py
# /// script
# requires-python = ">=3.12"
# dependencies = [
#     "genaiclass",
#     "agent-framework-core==1.19.0",
#     "agent-framework-openai==1.14.4",
#     "agent-framework-orchestrations==1.2.0",
# ]
#
# [tool.uv.sources]
# genaiclass = { path = "../../../shared/python", editable = true }
# ///
# %% [markdown]
# # 03 - Concurrent fan-out / fan-in
#
# | | |
# |---|---|
# | **Pattern** | concurrent: the same input goes to N agents at once, results are aggregated |
# | **Communication** | **broadcast** in, **aggregate** out - the agents never see each other |
# | **Use when** | independent perspectives: review panels, multi-source research, voting, ensembles |
# | **Watch out** | N agents = N times the cost; the aggregator is where quality is decided |
#
# ```bash
# uv run 03_maf_concurrent.py --profile mock
# ```

# %%
import asyncio

from agent_framework import Agent
from agent_framework.orchestrations import ConcurrentBuilder

from genaiclass import banner, get_profile, maf_client

profile = get_profile()
client = maf_client(profile)
print(banner(profile))

# %% [markdown]
# ## A review panel
#
# Isolation is the feature. The skeptic cannot be talked round by the optimist,
# because neither ever sees the other's answer - unlike a group chat (sample 05).

# %%
panel = [
    Agent(client, "You are an optimist. Give the strongest case FOR the idea in two sentences.",
          name="optimist"),
    Agent(client, "You are a skeptic. Give the strongest case AGAINST the idea in two sentences.",
          name="skeptic"),
    Agent(client, "You are a finance lead. Estimate cost and payback in two sentences.",
          name="finance"),
]
IDEA = "Replace our support hotline with an AI agent next quarter."


# %% [markdown]
# ## 1. Default aggregation
#
# Out of the box the results are merged into one response, one message per
# agent. Nothing is summarised - you get the raw panel.

# %%
async def default_aggregation() -> None:
    result = await ConcurrentBuilder(participants=panel).build().run(IDEA)
    for message in result.get_outputs()[-1].messages:
        print(f"  {message.author_name:<9} {message.text[:100]}")


# %% [markdown]
# ## 2. Your own aggregator
#
# The fan-in step is ordinary code. Here: a judge agent reads all three and
# decides. It could just as well be a vote count, a max-score pick, or a merge -
# this is the "reduce" of map/reduce, and it deserves as much design as the agents.

# %%
judge = Agent(client, "You are the decision maker. Weigh the panel and give a go/no-go in one line.",
              name="judge")


async def judged(results) -> str:
    panel_text = "\n".join(f"{r.executor_id}: {r.agent_response.text}" for r in results)
    verdict = await judge.run(f"Idea: {IDEA}\n\nPanel:\n{panel_text}")
    return verdict.text


async def custom_aggregation() -> None:
    workflow = ConcurrentBuilder(participants=panel).with_aggregator(judged).build()
    result = await workflow.run(IDEA)
    print(f"  verdict: {result.get_outputs()[-1]}")


async def main() -> None:
    print("--- 1. default aggregation (the raw panel)")
    await default_aggregation()
    print("\n--- 2. custom aggregator (a judge decides)")
    await custom_aggregation()


asyncio.run(main())

File: demos/samples/03-agents/csharp/03_maf_concurrent.cs
#:project ../../../shared/csharp/GenAIClass/GenAIClass.csproj
#:package Microsoft.Agents.AI@1.22.0
#:package Microsoft.Agents.AI.OpenAI@1.22.0
#:package Microsoft.Agents.AI.Workflows@1.22.0
#:property PublishAot=false

// %% [markdown]
// # 03 - Concurrent fan-out / fan-in (C#)
//
// | | |
// |---|---|
// | **Pattern** | concurrent: the same input goes to N agents at once, results are aggregated |
// | **Communication** | **broadcast** in, **aggregate** out - the agents never see each other |
// | **Use when** | independent perspectives: review panels, multi-source research, voting |
// | **Watch out** | N agents = N times the cost; the aggregator is where quality is decided |
//
// Twin of `python/03_maf_concurrent.py`.
//
// ```bash
// dotnet run 03_maf_concurrent.cs -- --profile mock
// ```

// %%
using GenAIClass;
using Microsoft.Agents.AI;
using Microsoft.Agents.AI.Workflows;
using Microsoft.Extensions.AI;
using OpenAI.Chat;
using ChatMessage = Microsoft.Extensions.AI.ChatMessage;

var profile = Providers.GetProfile(args: args);
Console.WriteLine(profile);
ChatClient chatClient = Providers.CreateClient(profile).GetChatClient(profile.Model);

const string Idea = "Replace our support hotline with an AI agent next quarter.";

AIAgent[] Panel() =>
[
    chatClient.AsAIAgent("You are an optimist. Give the strongest case FOR the idea in two sentences.", "optimist"),
    chatClient.AsAIAgent("You are a skeptic. Give the strongest case AGAINST the idea in two sentences.", "skeptic"),
    chatClient.AsAIAgent("You are a finance lead. Estimate cost and payback in two sentences.", "finance"),
];

// %% [markdown]
// ## The aggregator is a function over everyone's answers
//
// `BuildConcurrent(agents, aggregator)`: the aggregator receives one message
// list per agent and returns the workflow's output. Here a judge agent decides;
// a vote count or a merge would be just as valid. Without an aggregator you get
// the raw panel back.

// %%
AIAgent judge = chatClient.AsAIAgent("You are the decision maker. Weigh the panel and give a go/no-go in one line.", "judge");

List<ChatMessage> Judge(IList<List<ChatMessage>> answers)
{
    var panel = string.Join("\n", answers.Select(a => a.LastOrDefault()).Where(m => m is not null)
                                         .Select(m => $"{m!.AuthorName}: {m.Text}"));
    // The aggregator is synchronous here; keep the judge call short.
    var verdict = judge.RunAsync($"Idea: {Idea}\n\nPanel:\n{panel}").GetAwaiter().GetResult();
    return [new ChatMessage(ChatRole.Assistant, verdict.Text) { AuthorName = "judge" }];
}

async Task RunAsync(Workflow workflow, string title)
{
    Console.WriteLine($"\n--- {title}");
    await using StreamingRun run = await InProcessExecution.RunStreamingAsync(
        workflow, new List<ChatMessage> { new(ChatRole.User, Idea) });
    await run.TrySendMessageAsync(new TurnToken(emitEvents: true));

    await foreach (WorkflowEvent evt in run.WatchStreamAsync())
    {
        // Trap (verified on 1.22): AgentResponseUpdateEvent *derives from*
        // WorkflowOutputEvent. Test the specific type first, or the first
        // streamed token matches "output", carries no message list, and the
        // loop exits early with the workflow still running.
        if (evt is AgentResponseUpdateEvent)
            continue;
        if (evt is WorkflowOutputEvent output)
        {
            foreach (var message in output.As<List<ChatMessage>>() ?? [])
                if (message.Role == ChatRole.Assistant)
                    Console.WriteLine($"  {message.AuthorName,-9} {message.Text[..Math.Min(100, message.Text.Length)]}");
            break;
        }
    }
}

await RunAsync(AgentWorkflowBuilder.BuildConcurrent(Panel()), "1. default aggregation (the raw panel)");
await RunAsync(AgentWorkflowBuilder.BuildConcurrent(Panel(), Judge), "2. custom aggregator (a judge decides)");
